src/transformer_deploy/t5_utils/conversion_utils.py

```diff
@@ -0,0 +1,681 @@
+    input_ids_shape = input_ids.shape[0]
+    input_id_shape = TensorRTShape(
+        min_shape=[input_ids_shape, 1],
+        optimal_shape=[input_ids_shape, 1],
+        max_shape=[input_ids_shape, 256],
+        input_name="input_ids",
+    )
+    encoder_hidden_states_shape = TensorRTShape(
+        min_shape=[input_ids_shape, 1, 512],
+        optimal_shape=[input_ids_shape, 10, 512],
+        max_shape=[input_ids_shape, 200, 512],
+        input_name="encoder_hidden_states",
+    )
+
+    final_seq_len = TensorRTShape(
+        min_shape=[1],
+        optimal_shape=[1],
+        max_shape=[1],
+        input_name="final_seq_len",
+    )
+
+    input_shapes = [input_id_shape, encoder_hidden_states_shape, final_seq_len]
+    for i in range(num_layers):
+        input_shapes.append(
+            TensorRTShape(
+                min_shape=[input_ids_shape, 8, 0, 64],
+                optimal_shape=[input_ids_shape, 8, 100, 64],
+                max_shape=[input_ids_shape, 8, 200, 64],
+                input_name=f"past_key_values.{i}.decoder.key",
+            )
+        )
+        input_shapes.append(
+            TensorRTShape(
+                min_shape=[input_ids_shape, 8, 0, 64],
+                optimal_shape=[input_ids_shape, 8, 100, 64],
+                max_shape=[input_ids_shape, 8, 200, 64],
+                input_name=f"past_key_values.{i}.decoder.value",
+            )
+        )
+        input_shapes.append(
+            TensorRTShape(
+                min_shape=[input_ids_shape, 8, 0, 64],
+                optimal_shape=[input_ids_shape, 8, 10, 64],
+                max_shape=[input_ids_shape, 8, 200, 64],
+                input_name=f"past_key_values.{i}.encoder.key",
+            )
+        )
+        input_shapes.append(
+            TensorRTShape(
+                min_shape=[input_ids_shape, 8, 0, 64],
+                optimal_shape=[input_ids_shape, 8, 10, 64],
+                max_shape=[input_ids_shape, 8, 200, 64],
+                input_name=f"past_key_values.{i}.encoder.value",
+            )
+        )
+
+    return input_shapes
+
+
+def onnx_to_tensorrt_model(
+    runtime, onnx_model_path, trt_logger, tensor_shapes, workspace_size, quantization, tensorrt_model_path
+) -> Callable[[Dict[str, torch.Tensor]], Dict[str, torch.Tensor]]:
+    try:
+        from tensorrt.tensorrt import ICudaEngine
+
+        from transformer_deploy.backends.trt_utils import build_engine, load_engine, save_engine
+
+    except ImportError:
+        raise ImportError(
+            "It seems that TensorRT is not yet installed. "
+            "It is required when you declare TensorRT backend."
+            "Please find installation instruction on "
+            "https://docs.nvidia.com/deeplearning/tensorrt/install-guide/index.html"
+        )
+    model_engine: ICudaEngine = build_engine(
+        runtime=runtime,
+        onnx_file_path=onnx_model_path,
+        logger=trt_logger,
+        min_shape=tensor_shapes[0],
```

Comment:
This was causing me issues when calling in in convert.py#L359, since `tensor_shapes=input_shape` which is a list of `TensorRTShape` instead of a list of `int`. So I tried to fix it by changing the `onnx_to_tensorrt_model` to :
```python
if len(tensor_shapes) == 3:  # this is the encoder for t5
        model_engine: ICudaEngine = build_engine(
            runtime=runtime,
            onnx_file_path=onnx_model_path,
            logger=trt_logger,
            min_shape=tensor_shapes[0],
            optimal_shape=tensor_shapes[1],
            max_shape=tensor_shapes[2],
            workspace_size=workspace_size * 1024 * 1024,
            fp16=not quantization,
            int8=quantization,
        )
    else:  # this is the decoder for t5 and tensor_shapes is already a list of TensorRTShape
        model_engine: ICudaEngine = build_engine(
            runtime=runtime,
            onnx_file_path=onnx_model_path,
            logger=trt_logger,
            input_shapes=tensor_shapes,
            shape_tensors=[tensor_shapes[2]],  # final_seq_len
            workspace_size=workspace_size * 1024 * 1024,
            fp16=not quantization,
            int8=quantization,
        )
```
but now I'm getting the following error:
```
[02/20/2023-09:16:00] [TRT] [V] Registering layer: /block.0/layer.0/SelfAttention/Div_1 for ONNX node: /block.0/layer.0/SelfAttention/Div_1
[02/20/2023-09:16:00] [TRT] [E] [elementWiseLayer.cpp::validate::34] Error Code 4: Internal Error (/block.0/layer.0/SelfAttention/Div_1: operation DIV has incompatible input types Float and Int32)
[02/20/2023-09:16:00] [TRT] [W] building engine. depending on model size this may take a while
[02/20/2023-09:16:00] [TRT] [E] 4: [network.cpp::validate::2738] Error Code 4: Internal Error (Network must have at least one output)
[02/20/2023-09:16:00] [TRT] [E] 2: [builder.cpp::buildSerializedNetwork::738] Error Code 2: Internal Error (Assertion engine != nullptr failed. )
Traceback (most recent call last):
  File "/usr/local/bin/convert_model", line 8, in <module>
    sys.exit(entrypoint())
  File "/usr/local/lib/python3.8/dist-packages/transformer_deploy/convert.py", line 572, in entrypoint
    main(commands=args)
  File "/usr/local/lib/python3.8/dist-packages/transformer_deploy/convert.py", line 359, in main
    tensorrt_decoder = onnx_to_tensorrt_model(
  File "/usr/local/lib/python3.8/dist-packages/transformer_deploy/t5_utils/conversion_utils.py", line 795, in onnx_to_tensorrt_model
    model_engine: ICudaEngine = build_engine(
  File "/usr/local/lib/python3.8/dist-packages/transformer_deploy/backends/trt_utils.py", line 206, in build_engine
    engine: ICudaEngine = runtime.deserialize_cuda_engine(trt_engine)
TypeError: deserialize_cuda_engine(): incompatible function arguments. The following argument types are supported:
    1. (self: tensorrt.tensorrt.Runtime, serialized_engine: buffer) -> tensorrt.tensorrt.ICudaEngine

Invoked with: <tensorrt.tensorrt.Runtime object at 0x7fd643e8eeb0>, None
free(): invalid pointer
```

Reply:
I switched to `nvidia-tensorrt==8.4.3.1` now the code is crashing at  `parser.parse(model=f.read(), path=onnx_file_path)` in `trt_utils.py#L176` trying to parse `t5-dec-if-node/model.onnx` with the following logs:

```
...
[03/05/2023-13:24:34] [TRT] [V] Registering layer: node_of_output_/Shape_output_0_OutputLayer_6442 for ONNX node: node_of_output_/Shape_output_0_OutputLayer
[03/05/2023-13:24:34] [TRT] [V] Registering layer: node_of_output_/Shape_output_0_OutputLayer_6443 for ONNX node: node_of_output_/Shape_output_0_OutputLayer
[03/05/2023-13:24:34] [TRT] [V] Registering layer: node_of_output_/Shape_output_0_OutputLayer_6444 for ONNX node: node_of_output_/Shape_output_0_OutputLayer
[03/05/2023-13:24:34] [TRT] [E] node_of_output_/Shape_output_0_OutputLayer: IIfConditionalOutputLayer inputs must have the same shape.
```